### ISSUE TITLE
Detect invalid media type URL

### ISSUE THREAD

**Issue by melchoyce** [xwp/wp-core-media-widgets#149]
From @celloexpressions:

> No warning or error when using the wrong type of URL for a given widget. I accidentally pasted a YouTube URL into the audio widget and there were no warnings or errors, but it then shows the link and "download file". There needs to be feedback about why it's not working for this situation or, even better, the widget should adapt to the appropriate widget type based on the content used. For people who frequently use YouTube for music, the audio widget seems to suggest that a YouTube URL would work but then it gives a very odd experience.

We should try to figure out if we can detect when someone is pasting in a URL for the wrong kind of media type. I'll design an error state and then we can explore more.

**Comment by obenland** [comment]
Dupe of #128?

**Comment by westonruter** [comment]
@obenland actually not. It's closely related, but distinct. The issue in #128 is regarding restricting media types during file _upload_. This issue is different in that it is restricting the URLs that can be supplied when embedding external media.

**Comment by westonruter** [comment]
@melchoyce see #157 for a work in progress.

**Comment by timmyc** [comment]
I tested out the PR and left some notes.  My feeling right now is if we are going to add an error state here - we should display a message around _why_ the URL entered is/might fail - and keep the insert button enabled.  Just seems like there are many scenarios where logic to determine if a URL is valid for a given widget could fail... perhaps the embed endpoint is down, or the resource being linked does not have a file extension for some reason.

Ultimately, I think the best thing we can do is try to ensure we always have a good preview rendered in the modal and also warn the user if we think there might be something wrong with what they have entered.

**Comment by melchoyce** [comment]
> I tested out the PR and left some notes. My feeling right now is if we are going to add an error state here - we should display a message around why the URL entered is/might fail

Agreed — here's a quick stab at what that could be: 

<img width="1252" alt="screen shot 2017-05-09 at 3 48 16 pm" 
src="https://cloud.githubusercontent.com/assets/2846578/25869628/099d8dc0-34cf-11e7-947e-874bb51e7032.png">

```
<div class="notice notice-error notice-alt">
    <p>Looks like you're trying to add a video file. Please link to an audio file instead.</p>
</div>
```

```
.media-frame-content .notice {
    margin-top: 16px;
}
```

That particular text might be too specific. Let me know what kind of information we have here so I can write up a couple error copy options.

**Comment by westonruter** [comment]
> My feeling right now is if we are going to add an error state here - we should display a message around why the URL entered is/might fail

Agreed. I didn't add that because I figured you'd want to design it 😄 

> That particular text might be too specific. Let me know what kind of information we have here so I can write up a couple error copy options.

Ideally the message could be more generic. Like an unrecognized file type (e.g. `.exe`) or an invalid file type (using an MP3 for a video), or an oEmbed that isn't for the expected type. If the messages could be generic then we could avoid having to tailor them for each type. At the very least, I think the messages should only mention the target media type and not also mention the type that was incorrectly provided. So like, “This doesn't appear to be a valid URL to an audio file.”

**Comment by melchoyce** [comment]
Maybe...:

> Looks like this isn't the correct kind of file. Please link to an [audio] file instead.

**Comment by westonruter** [comment]
@melchoyce now implemented in #157 (which clearly has some overlap with #142):

![image](https://cloud.githubusercontent.com/assets/134745/25886025/956cc23a-3510-11e7-972a-b633f1522d40.png)

**Comment by melchoyce** [comment]
Hmm, unsure about that copy — lots of "ors," and kind of hard to parse now with the added complexity. Is the concern that "video file" excludes YouTube, Vimeo, and other streaming services?

**Comment by westonruter** [comment]
@melchoyce yes. For video specifically, it seems we need something to the effect of a URL to an actual file or a URL to a video on a streaming service.

**Comment by melchoyce** [comment]
Maybe:

> This doesn’t look like the right kind of file. Please link to a video file or stream instead.

Props @ranh @kjellr for copy editing.

**Comment by melchoyce** [comment]
(And that works for audio, too!)

**Comment by westonruter** [comment]
@melchoyce for video streams, since we only support YouTube and Vimeo, should these not be explicitly called out? I can imagine someone attempting to link to a Brightcove video (or something) but being confused as to why it is not working.

**Comment by melchoyce** [comment]
We support embeds in core from other video services (there's a ton here: https://codex.wordpress.org/Embeds#Okay.2C_So_What_Sites_Can_I_Embed_From.3F). Are those all excluded from the widget? Just clarifying.

**Comment by timmyc** [comment]
> Are those all excluded from the widget?

Once we have the oembed/proxy endpoint merged and in use here, all supported core embed video services _should_ work 🤞

**Comment by melchoyce** [comment]
In that case, let's be vague. If we can't get that endpoint merged in time, we'll re-asses.

**Comment by westonruter** [comment]
> Once we have the oembed/proxy endpoint merged and in use here, all supported core embed video services should work 🤞

@timmyc Unfortunately not. The video shortcode explicitly only supports YouTube and Vimeo: https://github.com/WordPress/wordpress-develop/blob/68320b3d9bee2d8146028c36dddd4efa56364083/src/wp-includes/media.php#L2489-L2502

This is because MediaElement.js is only able to wrap the players for these two services, at the time of writing the shortcode at any case.

**Comment by westonruter** [comment]
In being vague, what about adding “supported” in the language:

> This doesn’t look like the right kind of file. Please link to a supported video file or stream instead.

Should we then take that further and list out the specific types that are supported (as in #172) while then also mention the video streams that are supported (YouTube and Vimeo)?

**Comment by melchoyce** [comment]
It might be info overload, but it would solve the question of "but what IS supported?" if someone's having issues.

> This doesn’t look like the right kind of file. Please link to a supported video file (.mp4, .m4v, .webm, .ogv, .wmv, .flv) or stream (YouTube, Vimeo) instead.

**Comment by westonruter** [comment]
@melchoyce how's this:

![image](https://cloud.githubusercontent.com/assets/134745/25923832/3fa000b8-3594-11e7-968b-d8da9fa4c384.png)

Also slightly adjusted the error message introduced in #172 so that it can be re-used in both places:

![image](https://cloud.githubusercontent.com/assets/134745/25923852/5d9ac850-3594-11e7-8bcc-7f84cdb0bac0.png)

Code is in PR #157.

**Comment by melchoyce** [comment]
Cool, as long as it also uses the CSS I mentioned in https://github.com/xwp/wp-core-media-widgets/pull/172#issuecomment-300611416.
